File: recursion/python/solutions/concat.py
"""Solution for concat."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def concat(lst1, lst2):
    """
    Concatenate two lists.

    concat(1 -> 2, 3 -> 4) => 1 -> 2 -> 3 -> 4
    concat((), 1 -> 2) => 1 -> 2
    concat(1 -> 2, ()) => 1 -> 2

    Note: Recurse on lst1, lst2 stays fixed.
    """
    if is_empty(lst1):
        return lst2

    first, rest = unprepend(lst1)

    return prepend(first, concat(rest, lst2))


if __name__ == '__main__':
    # Test cases
    assert to_python_list(concat(Node.from_list([1, 2]), Node.from_list([3, 4]))) == [1, 2, 3, 4]
    assert to_python_list(concat(EMPTY_LIST, Node.from_list([1, 2]))) == [1, 2]
    assert to_python_list(concat(Node.from_list([1, 2]), EMPTY_LIST)) == [1, 2]
    assert to_python_list(concat(EMPTY_LIST, EMPTY_LIST)) == []

    print("All tests passed!")

File: recursion/javascript/solutions/concat.js
/**
 * Solution for concat.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function concat(list1, list2) {
  if (isEmpty(list1)) {
    return list2;
  }

  let [first, rest] = unprepend(list1);

  return prepend(first, concat(rest, list2));
}

if (require.main === module) {
  console.assert(JSON.stringify(toArray(concat(Node.fromArray([1, 2]), Node.fromArray([3, 4])))) === '[1,2,3,4]');
  console.assert(JSON.stringify(toArray(concat(EMPTY_LIST, Node.fromArray([1, 2])))) === '[1,2]');
  console.assert(JSON.stringify(toArray(concat(Node.fromArray([1, 2]), EMPTY_LIST))) === '[1,2]');
  console.assert(JSON.stringify(toArray(concat(EMPTY_LIST, EMPTY_LIST))) === '[]');

  console.log('All tests passed!');
}

module.exports = { concat };
